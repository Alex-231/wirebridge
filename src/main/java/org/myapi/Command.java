package org.myapi;

import com.spun.util.ObjectUtils;
import org.myapi.models.ConnectionDetails;
import org.myapi.models.DatabaseConfig;

import java.sql.*;
import java.util.HashMap;
import java.util.Map;

public class Command {

    private ConnectionDetails config;
    private HashMap<String, DatabaseConfig> databases;

    public Command(ConnectionDetails connectionDetails, HashMap<String, DatabaseConfig> databases) {
        config = connectionDetails;
        this.databases = databases;
    }

    public QueryResult call(HashMap<String, String> params) {
//        Create DB connection
        DatabaseConfig database = getDatabaseFor(config.getSql().getDatabase());

        Query sql = null;

        try {
            Connection connection = DriverManager.getConnection(database.getUrl(), database.getUsername(), database.getPassword());

//        Mix params in SQL
            sql = config.getSql().createSqlFor(params);
            PreparedStatement preparedStatement = connection.prepareStatement(sql.getQueryText());
            preparedStatement.setString(1, params.get(sql.getParameterNames()[0]));
            boolean returnResultSet = preparedStatement.execute();

            return new QueryResult(preparedStatement.getResultSet(), preparedStatement.getUpdateCount());
        } catch (SQLException e) {
//            Call SQL
            System.out.println("SQL Error: " + sql);
            ObjectUtils.throwAsError(e);
        }

        return null;
    }

    private DatabaseConfig getDatabaseFor(String database) {
        return databases.get(database);
    }
}
