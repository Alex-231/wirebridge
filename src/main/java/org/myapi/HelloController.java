package org.myapi;

import com.spun.util.io.FileUtils;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import javax.servlet.http.HttpServletRequest;
import java.util.Map;

@RestController
@EnableAutoConfiguration
public class HelloController {

    @RequestMapping("/**")
    String home(HttpServletRequest request, @RequestParam Map<String,String> allRequestParams) {
        MyAPIStartup myApi = new MyAPIStartup();
        myApi.loadJson(FileUtils.readFromClassPath(getClass(), "../../add_table.json"));
        myApi.loadJson(FileUtils.readFromClassPath(getClass(), "../../query_table.json"));
        myApi.loadJson(FileUtils.readFromClassPath(getClass(), "../../config.json"));
        QueryResult queryResult = myApi.call(request.getMethod(), request.getRequestURI(), allRequestParams);

        return queryResult.toString();
    }

    public static void main(String[] args) throws Exception {
        SpringApplication.run(HelloController.class, args);
    }

}
