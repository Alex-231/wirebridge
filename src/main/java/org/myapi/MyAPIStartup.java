package org.myapi;

import com.google.gson.Gson;
import com.spun.util.io.FileUtils;
import org.myapi.models.ConnectionDetails;
import org.myapi.models.DatabaseConfig;
import org.myapi.models.Request;

import java.io.IOException;
import java.sql.ResultSet;
import java.util.HashMap;

public class MyAPIStartup {
    private HashMap<String, Command> call = new HashMap<String, Command>();
    private HashMap<String, DatabaseConfig> databases = new HashMap<String, DatabaseConfig>();

    public QueryResult call(String httpMethod, String path, HashMap<String, String> params) {
        Command command = call.get(httpMethod + ":" + path);
        return command.call(params);
    }

    public void loadJsonFromFile(String configFile) throws IOException {
        loadJson(FileUtils.readFile(configFile));
    }

    public void loadJson(String json) {
        Gson g = new Gson();
        ConnectionDetails connectionDetails = g.fromJson(json, ConnectionDetails.class);

        Command command = new Command(connectionDetails, databases);
        addDatabases(connectionDetails);
        addRequest(connectionDetails, command);
    }

    private void addRequest(ConnectionDetails connectionDetails, Command command) {
        Request request = connectionDetails.getRequest();

        if(request != null){
            call.put(request.getMethod() + ":" + request.getPath(),command);
        }
    }

    private void addDatabases(ConnectionDetails connectionDetails) {
        DatabaseConfig database = connectionDetails.getDatabase();

        if(database != null){
            databases.put(database.getName(), database);
        }
    }
}
