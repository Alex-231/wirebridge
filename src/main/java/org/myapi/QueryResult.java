package org.myapi;

import java.sql.ResultSet;

public class QueryResult {

    public final ResultSet resultSet;
    public final int updateCount;

    public QueryResult(ResultSet resultSet, int updateCount) {

        this.resultSet = resultSet;
        this.updateCount = updateCount;
    }

    @Override
    public String toString() {
        return "QueryResult{" +
                "resultSet=" + resultSet +
                ", updateCount=" + updateCount +
                '}';
    }
}
