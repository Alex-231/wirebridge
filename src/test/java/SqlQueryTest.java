import com.spun.util.io.FileUtils;
import org.approvaltests.Approvals;
import org.junit.Test;
import org.myapi.MyAPIStartup;
import org.myapi.QueryResult;

import java.io.IOException;
import java.util.HashMap;


public class SqlQueryTest{

    @Test
    public void testInsertIntoDB() throws IOException {
        HashMap<String,String> parameters = new HashMap<String, String>(){{
            put("title", "Cool API Demo!");
        }};

        verifyQuery("addtable.json", "POST", "/talk", parameters);
    }

    private void verifyQuery(String configFile, String httpMethod, String path, HashMap<String, String> params) throws IOException {
        MyAPIStartup myApi = new MyAPIStartup();
        myApi.loadJson(FileUtils.readFromClassPath(getClass(), "config.json"));
        myApi.loadJson(FileUtils.readFromClassPath(getClass(), configFile));
        QueryResult queryResult = myApi.call(httpMethod, path, params);

        Approvals.verify(queryResult.toString());
    }

}
