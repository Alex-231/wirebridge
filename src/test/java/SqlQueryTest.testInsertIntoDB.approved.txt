QueryResult{resultSet=null, updateCount=1}
